File: platforms/android/native-netkit/src/main/java/com/aifirst/nativenetkit/NativeNetKit.kt
package com.aifirst.nativenetkit

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.channels.awaitClose
import kotlinx.coroutines.channels.trySendBlocking
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.callbackFlow
import kotlinx.coroutines.withContext
import okhttp3.Call
import okhttp3.MediaType.Companion.toMediaTypeOrNull
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.Response
import okhttp3.sse.EventSource
import okhttp3.sse.EventSourceListener
import okhttp3.sse.EventSources
import java.io.IOException
import kotlin.system.measureTimeMillis

data class NativeRequest(
    val method: String = "GET",
    val url: String,
    val headers: Map<String, List<String>> = emptyMap(),
    val body: ByteArray? = null,
)

data class NativeResponse(
    val statusCode: Int,
    val headers: Map<String, List<String>> = emptyMap(),
    val body: ByteArray = ByteArray(0),
    val elapsedMilliseconds: Long? = null,
)

data class NativeSseEvent(
    val id: String? = null,
    val type: String? = null,
    val data: String,
)

enum class NativeNetworkErrorCode {
    INVALID_REQUEST,
    TRANSPORT_FAILURE,
    CANCELLED,
    UNKNOWN,
}

class NativeNetworkException(
    val code: NativeNetworkErrorCode,
    override val message: String,
    val rawDescription: String? = null,
    cause: Throwable? = null,
) : Exception(message, cause)

fun interface NativeHttpEngine {
    suspend fun execute(request: NativeRequest): NativeResponse
}

fun interface NativeSseEngine {
    fun stream(request: NativeRequest): Flow<NativeSseEvent>
}

class OkHttpNativeHttpEngine(
    private val callFactory: Call.Factory = OkHttpClient(),
) : NativeHttpEngine {
    override suspend fun execute(request: NativeRequest): NativeResponse = withContext(Dispatchers.IO) {
        val okhttpRequest = try {
            request.toOkHttpRequest()
        } catch (error: IllegalArgumentException) {
            throw NativeNetworkException(
                code = NativeNetworkErrorCode.INVALID_REQUEST,
                message = "Invalid request",
                rawDescription = error.message,
                cause = error,
            )
        }

        var responseBytes = ByteArray(0)
        var statusCode = -1
        var headers: Map<String, List<String>> = emptyMap()
        val elapsed = try {
            measureTimeMillis {
                callFactory.newCall(okhttpRequest).execute().use { response ->
                    statusCode = response.code
                    headers = response.headers.toMultimap()
                    responseBytes = response.body?.bytes() ?: ByteArray(0)
                }
            }
        } catch (error: IOException) {
            throw NativeNetworkException(
                code = NativeNetworkErrorCode.TRANSPORT_FAILURE,
                message = "OkHttp request failed",
                rawDescription = error.message,
                cause = error,
            )
        }

        NativeResponse(
            statusCode = statusCode,
            headers = headers,
            body = responseBytes,
            elapsedMilliseconds = elapsed,
        )
    }
}

class OkHttpNativeSseEngine(
    private val eventSourceFactory: EventSource.Factory = EventSources.createFactory(OkHttpClient()),
) : NativeSseEngine {
    override fun stream(request: NativeRequest): Flow<NativeSseEvent> = callbackFlow {
        val okhttpRequest = try {
            request.toOkHttpRequest(defaultAccept = "text/event-stream")
        } catch (error: IllegalArgumentException) {
            close(
                NativeNetworkException(
                    code = NativeNetworkErrorCode.INVALID_REQUEST,
                    message = "Invalid request",
                    rawDescription = error.message,
                    cause = error,
                ),
            )
            return@callbackFlow
        }

        val eventSource = eventSourceFactory.newEventSource(
            okhttpRequest,
            object : EventSourceListener() {
                override fun onEvent(eventSource: EventSource, id: String?, type: String?, data: String) {
                    val result = trySendBlocking(NativeSseEvent(id = id, type = type, data = data))
                    if (result.isFailure) {
                        eventSource.cancel()
                    }
                }

                override fun onClosed(eventSource: EventSource) {
                    close()
                }

                override fun onFailure(eventSource: EventSource, t: Throwable?, response: Response?) {
                    close(
                        NativeNetworkException(
                            code = NativeNetworkErrorCode.TRANSPORT_FAILURE,
                            message = "OkHttp SSE stream failed",
                            rawDescription = t?.message ?: response?.let { "HTTP ${it.code}" },
                            cause = t,
                        ),
                    )
                }
            },
        )

        awaitClose { eventSource.cancel() }
    }
}

class NativeNetClient(
    private val engine: NativeHttpEngine = OkHttpNativeHttpEngine(),
    private val sseEngine: NativeSseEngine = OkHttpNativeSseEngine(),
) {
    constructor(engine: NativeHttpEngine) : this(engine, OkHttpNativeSseEngine())

    suspend fun execute(request: NativeRequest): NativeResponse {
        return engine.execute(request)
    }

    suspend fun get(url: String, headers: Map<String, List<String>> = emptyMap()): NativeResponse {
        return execute(NativeRequest(method = "GET", url = url, headers = headers))
    }

    fun stream(request: NativeRequest): Flow<NativeSseEvent> {
        return sseEngine.stream(request)
    }

    fun stream(url: String, headers: Map<String, List<String>> = emptyMap()): Flow<NativeSseEvent> {
        return stream(NativeRequest(method = "GET", url = url, headers = headers))
    }
}

private fun NativeRequest.toOkHttpRequest(defaultAccept: String? = null): Request {
    val builder = Request.Builder().url(url)

    headers.forEach { (name, values) ->
        values.forEach { value -> builder.addHeader(name, value) }
    }

    if (defaultAccept != null && headers.keys.none { it.equals("Accept", ignoreCase = true) }) {
        builder.addHeader("Accept", defaultAccept)
    }

    val requestBody = body?.toRequestBody("application/octet-stream".toMediaTypeOrNull())
    builder.method(method, requestBody)
    return builder.build()
}

File: platforms/ios/Sources/NativeNetKit/NativeNetKit.swift
import Foundation

public struct NativeRequest: Sendable {
    public var method: String
    public var url: URL
    public var headers: [String: [String]]
    public var body: Data?

    public init(
        method: String = "GET",
        url: URL,
        headers: [String: [String]] = [:],
        body: Data? = nil
    ) {
        self.method = method
        self.url = url
        self.headers = headers
        self.body = body
    }
}

public struct NativeResponse: Sendable, Equatable {
    public var statusCode: Int
    public var headers: [String: [String]]
    public var body: Data
    public var elapsedMilliseconds: Int64?

    public init(
        statusCode: Int,
        headers: [String: [String]] = [:],
        body: Data = Data(),
        elapsedMilliseconds: Int64? = nil
    ) {
        self.statusCode = statusCode
        self.headers = headers
        self.body = body
        self.elapsedMilliseconds = elapsedMilliseconds
    }
}

public enum NativeNetworkErrorCode: String, Sendable, Equatable {
    case invalidRequest
    case transportFailure
    case cancelled
    case unknown
}

public struct NativeNetworkError: Error, Sendable, Equatable {
    public var code: NativeNetworkErrorCode
    public var message: String
    public var rawDescription: String?

    public init(
        code: NativeNetworkErrorCode,
        message: String,
        rawDescription: String? = nil
    ) {
        self.code = code
        self.message = message
        self.rawDescription = rawDescription
    }
}

public protocol NativeHttpEngine: Sendable {
    func execute(_ request: NativeRequest) async throws -> NativeResponse
}

public final class URLSessionNativeHttpEngine: NativeHttpEngine {
    private let session: URLSession

    public init(session: URLSession = .shared) {
        self.session = session
    }

    public func execute(_ request: NativeRequest) async throws -> NativeResponse {
        var urlRequest = URLRequest(url: request.url)
        urlRequest.httpMethod = request.method
        urlRequest.httpBody = request.body

        for (name, values) in request.headers {
            for value in values {
                urlRequest.addValue(value, forHTTPHeaderField: name)
            }
        }

        let start = Date()

        do {
            let (data, response) = try await session.data(for: urlRequest)
            guard let httpResponse = response as? HTTPURLResponse else {
                throw NativeNetworkError(
                    code: .transportFailure,
                    message: "Response was not an HTTP response"
                )
            }

            let elapsedMilliseconds = Int64(Date().timeIntervalSince(start) * 1_000)
            return NativeResponse(
                statusCode: httpResponse.statusCode,
                headers: httpResponse.allHeaderFields.reduce(into: [:]) { result, item in
                    guard let key = item.key as? String else { return }
                    result[key, default: []].append(String(describing: item.value))
                },
                body: data,
                elapsedMilliseconds: elapsedMilliseconds
            )
        } catch is CancellationError {
            throw NativeNetworkError(code: .cancelled, message: "Request was cancelled")
        } catch let error as NativeNetworkError {
            throw error
        } catch {
            throw NativeNetworkError(
                code: .transportFailure,
                message: "URLSession request failed",
                rawDescription: String(describing: error)
            )
        }
    }
}

public final class NativeNetClient: Sendable {
    private let engine: NativeHttpEngine

    public init(engine: NativeHttpEngine = URLSessionNativeHttpEngine()) {
        self.engine = engine
    }

    public func execute(_ request: NativeRequest) async throws -> NativeResponse {
        try await engine.execute(request)
    }

    public func get(_ url: URL, headers: [String: [String]] = [:]) async throws -> NativeResponse {
        try await execute(NativeRequest(method: "GET", url: url, headers: headers))
    }
}
